core/src/main/scala/org/apache/spark/sql/execution/columnar/encoding/DictionaryEncoding.scala

```diff
@@ -106,96 +118,46 @@ abstract class DictionaryDecoderBase
       case _ => throw new UnsupportedOperationException(
         s"DictionaryEncoding not supported for $dataType")
     }
-    baseCursor = position
     position
   }
 
-  private[sql] def initializeBeforeFinish(dictionaryBytes: AnyRef,
-      dictionaryCursor: Long, dictionaryLen: Int, cursor: Long, dataType: DataType): Unit = {
-    initializeDictionary(dictionaryBytes, dictionaryCursor,
-      dictionaryLen, dictionaryLen, dataType)
-    // base cursor will start at the given cursor since dictionary is separate
-    baseCursor = cursor
-  }
-
-  override def nextUTF8String(columnBytes: AnyRef, cursor: Long): Long =
-    cursor + 2
-
-  override def absoluteUTF8String(columnBytes: AnyRef, position: Int): Long = {
-    // TODO: PERF: optimize for local index access case by filling
-    // in the dictionary array lazily on access
-    baseCursor + ((position - numNullsUntilPosition(columnBytes, position)) << 1)
-  }
-
-  override def readUTF8String(columnBytes: AnyRef, cursor: Long): UTF8String =
-    stringDictionary.getString(columnBytes, ColumnEncoding.readShort(columnBytes, cursor))
+  // TODO: PERF: optimize for local index access case by filling
+  // in the dictionary array lazily on access
+  override def readUTF8String(columnBytes: AnyRef, nonNullPosition: Int): UTF8String =
+  stringDictionary.getString(columnBytes,
```

Comment:
For a join query on string column with a big dictionary (say when String is kind of primary key), what is the overhead of creation of objects in every call? Since we are trying to solve the problem of promotion of dictionary objects, another approach could be to divide the dictionary into multiple sub dictionaries. Say for e.g. after every 1000 entries, rollover to a new dictionary. Each dictionary has a start and end ordinals of columns which can be used by the decoder. Alternatively, we can put the dictionary id in addition to dictionary index inside every entry. This approach could be generic so that it can be extended when we move to a mechanism of having a common dictionary across cached batches. Theoretically, the queries should have little impact.